numpy/__init__.pyi

```diff
@@ -2578,12 +2581,18 @@ class ndarray(_ArrayOrScalarCommon, Generic[_ShapeType, _DType_co]):
 
     def __dlpack_device__(self) -> tuple[int, L[0]]: ...
 
-    def __array_namespace__(self, *, api_version: str | None = ...) -> Any: ...
-
-    def to_device(self, device: L["cpu"], /, *, stream: None | int | Any = ...) -> NDArray[Any]: ...
-
-    @property
-    def device(self) -> L["cpu"]: ...
+    @overload
+    def to_device(self: NDArray[_UnknownType], device: L["cpu"], /, *, stream: None | int | Any = ...) -> NDArray[Any]: ...
+    @overload
+    def to_device(self: NDArray[np.bool], device: L["cpu"], /, *, stream: None | int | Any = ...) -> NDArray[np.bool]: ...
+    @overload
+    def to_device(self: NDArray[_NumberType], device: L["cpu"], /, *, stream: None | int | Any = ...) -> NDArray[_NumberType]: ...
+    @overload
+    def to_device(self: NDArray[timedelta64], device: L["cpu"], /, *, stream: None | int | Any = ...) -> NDArray[timedelta64]: ...
+    @overload
+    def to_device(self: NDArray[object_], device: L["cpu"], /, *, stream: None | int | Any = ...) -> NDArray[object_]: ...
+    @overload
+    def to_device(self: NDArray[Any], device: L["cpu"], /, *, stream: None | int | Any = ...) -> NDArray[Any]: ...
```

Comment:
Do we really need all these overloads, isn't the last one enough?

Reply:
With these overloads we can have more specialized typing, like:
```python
assert_type(AR_f8.to_device("cpu"), npt.NDArray[np.float64])
assert_type(AR_i8.to_device("cpu"), npt.NDArray[np.int64])
```
Similar overloads are present for `__abs__` or `__invert__` methods.

Without these overloads the typing result of the `to_device()` call will always be `NDArray[Any]`.

Reply:
OK, but it all seems very verbose to me. Isn't there a more concise way to say that the output dtype is the same as the input dtype?

Reply:
Ah, right. I checked it and it's possible - I reduced it to two overloads: one that says input dtype is the same as output, and one that returns `Any`. I also expanded typing tests for `to_device`.

Reply:
Thanks. Curious: what does `_SCT` stand for?